linux-setup.sh

```diff
@@ -0,0 +1,273 @@
+#!/bin/bash -e
+
+# This is a script to perform build and install processes for the `haiku-format` tool on a
+# Linux platform. Run the script with no arguments to get usage. See the accompanying
+# `linux.md` file for instructions on general use.
+
+set -e
+set -o pipefail
+
+script_bin=$0
+build_dir="llvm-project/build"
+install_dir="/opt/haiku-format"
+llvm_project=llvm-project
+
+function hf_usage()
+{
+	cat <<EOF
+${script_bin}
+
+This script is designed to build the 'haiku-format' tool and install it on an
+'apt'-based Linux system.
+
+${script_bin} build
+${script_bin} clean
+sudo ${script_bin} install
+sudo ${script_bin} uninstall
+
+EOF
+
+	exit 1
+}
+
+function hf_confirm_yes_no()
+{
+	while true; do
+		read -p "$* [y/n]: " yn
+		case $yn in
+			[Yy])
+				return 0
+				;;
+			[Nn])
+				echo "aborted" 1>&2
+				return 1
+				;;
+		esac
+	done
+}
+
+# Detects absent Debian packages that are required for build and install. If there are
+# any absent then it will error to the user with the command they are required to run
+# to install the missing packages.
+
+function hf_install_dependencies_linux_gnu_apt()
+{
+	local depends="lld clang cmake ninja-build"
+	local missing_depends=()
+
+	for d in $depends; do
+		if ! dpkg -s "${d}" &> /dev/null; then
+			missing_depends+=("${d}")
+		fi
+	done
+
+	if [ "${#missing_depends[@]}" -gt 0 ]; then
+		echo "ensure that the necessary dependencies are installed before running this script;"
+		echo "sudo apt install ${missing_depends[*]}"
+		exit 1
+	fi
+}
+
+# A number of source files have to be downloaded for the build process to work. This
+# function will pull those down and then un-pack them into the directory where the
+# build will be undertaken.
+
+function hf_download_sources()
+{
+	if [ $# != 1 ]; then
+		echo "missing llvm version to download sources"
+		exit 1
+	fi
+
+	local llvm_version="$1"
+	local assets="clang cmake llvm third-party"
+	local tar_suffix="${llvm_version}.src.tar.xz"
+	local llvm_base_url="https://github.com/llvm/llvm-project/releases/download/llvmorg-${llvm_version}"
+	local tarball
+
+	for a in $assets; do
+		tarball="${a}-${tar_suffix}"
+		if [ -e "${tarball}" ]; then
+			echo "file [${tarball}] exists - can skip download"
+		else
+			echo "will download [${tarball}]"
+			wget -N "${llvm_base_url}/${tarball}"
+		fi
+	done
+
+	mkdir -v "${llvm_project}"
+
+	for a in $assets; do
+		tarball="${a}-${tar_suffix}"
+		mkdir -v "${llvm_project}/${a}"
+		echo -n "will extract ${a}"
+		tar -xf "${tarball}" -C "${llvm_project}/${a}" --strip-components=1 --checkpoint=.1000
+		echo
+	done
+}
+
+# This function will download and unpack the sources and then perform the build. It will
+# not proceed if there is already a build product present in the build location.
+
+function hf_build()
+{
+	local llvm_version
+
+	llvm_version="$(ls -v v*.diff | tail -1 | sed -E 's/v([0-9]+(\.[0-9]+){2})\.diff/\1/')"
+
+	if [ -z "$llvm_version" ]; then
+		exit 1
+	fi
```

Comment:
My bad as I didn't see the check on line 122 below, which already covers this empty string check. Will delete it before merging.
```suggestion
```